File: problems/42. Trapping Rain Water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0
        result = 0
        lMax = 0
        rMaxes = [0]*len(height)
        rMaxes[-1] = height[-1]
        for i in range(len(height) - 2, 0, -1):
            rMaxes[i] = max(rMaxes[i+1], height[i+1])

        for i in range(1, len(height)-1):
            lMax = max(height[i-1], lMax)
            rMax = rMaxes[i]
            waterHeight = min(lMax, rMax) - height[i]
            result += waterHeight if waterHeight > 0 else 0

        return result

File: problems/42. Trapping Rain Water/solution.cpp
class Solution
{
public:
    int trap(vector<int> &height)
    {
        if (height.empty())
            return 0;
        int result(0), lMax(0);
        vector<int> rMax(height.size(), 0);
        rMax[rMax.size() - 1] = height[height.size() - 1];
        for (int i = height.size() - 2; i >= 0; i--)
            rMax[i] = max(height[i], rMax[i + 1]);
        for (int i = 0; i < height.size(); i++)
        {
            lMax = max(height[i], lMax);
            result += min(rMax[i], lMax) - height[i];
        }
        return result;
    }
};
